### DISCUSSION TITLE
Pre-filled values are ignored when server side rendering

### DISCUSSION THREAD

**Discussion by pooyarm** [react-hook-form/react-hook-form#7576] [Q&A] [answered]
### The problem
Recently I faced a weird problem! Our product is based on nextjs, on our login page we have two fields, email and password, when the user's browser pre-fills the values and user tries to submit the form without touching the fields, the validation error happens, actually react-hook-form considers the fields as empty! I've checked it on Firefox.

#### Here is the example:
https://codesandbox.io/s/lucid-almeida-x3bld?file=/pages/index.js

#### And here is the video of what happens:

https://user-images.githubusercontent.com/37490/149973901-af846a12-be7c-4203-a20f-d1a402e849c1.mp4

### Workaround:
If we force the form to be rendered on client-side it'll be fixed, but I don't like this solution.

**Comment by Moshyfawn** [comment]
It's working if you register your inputs via [`register` form method](https://react-hook-form.com/api/useform/register#main), huh.. See [this CSB](https://codesandbox.io/s/thf-autofill-controlled-issue-hstzw)

**Reply by pooyarm** [reply]
I think we shouldn't be restricted to using `register` for such a case. btw, your csb link doesn't work for me.

**Reply by Moshyfawn** [reply]
Oh, totally, you're right! I've created an issue on the matter. See #7578
Sorry for a broken CSB. I've updated my answer with a working one. This is the new CSB: [link](https://codesandbox.io/s/thf-autofill-controlled-issue-hstzw)

**Comment by Moshyfawn** [comment, marked as answer]
The issue is out of scope of RHF. See [this comment](https://github.com/react-hook-form/react-hook-form/issues/7578#issuecomment-1015886202) on #7578.

It looks like your best bet is to render the form that might rely on the autofill on the client or try using uncontrolled inputs with `register`

**Reply by pooyarm** [reply]
I don't understand the reason why it's out of scope.

**Reply by Moshyfawn** [reply]
Because it's not RHF managing the state magically. The controlled input state is handled by React. There're numerous issues and discussions about the issue, since every browser implements their autofill differently; some fire the `onChange` event, some don't etc. See facebook/react#21459, facebook/react#15739 for context. There's a bunch more similar issues across React's repo and outside of it

**Reply by Moshyfawn** [reply]
See other repos moving such issues to "the correct repo": mui-org/material-ui#16943, ant-design/ant-design#21232

**Comment by Moshyfawn** [comment]
Here's another hacky solution I can propose. See [the same CSB](https://codesandbox.io/s/thf-autofill-controlled-issue-hstzw)
Basically, you query the DOM input value and set a non-empty value via `setValue` at the start. You might wanna tweak the solution
```js
const hacky_handleWeirdAutofillIssue = useCallback(
  (fieldName, fieldValue) => {
    setValue(fieldName, fieldValue);
  },
  [setValue]
);

useEffect(() => {
  const unregisteredDomAutofillValue = document.querySelector(
    "input[name='email']"
  )?.value;
  if (unregisteredDomAutofillValue) {
    hacky_handleWeirdAutofillIssue();
  }
}, [hacky_handleWeirdAutofillIssue]);
```

**Reply by pooyarm** [reply]
I've also did the same approach, had to query elements and set their values on an effect. but another thing that I expected from RHF was to be able to get fields and their refs somewhere in `useForm` object. I think such a feature can be helpful for this kind of situations.

**Reply by Moshyfawn** [reply]
You can do

```js
const { field: { ref, onChange, ...fieldProps } } = useController({ name })
const inputRef = useRef(null);

useEffect(() => {
  if (inputRef.current) {
    ref(inputRef.current);
    // INFO: your can pass the DOM input element value **onChange**,  **setValue** or completely **reset** your form
    // with the auto-filled values
    if (inputRef.current.value) {
      onChange(inputRef.current.value);
    }
  }
}, [ref, name, onChange]);

return  <input {...fieldProps} ref={inputRef}  onChange={onChange} />

```
_P.S. I've omitted some of the values for readability purpose_

See [this CSB](https://codesandbox.io/s/thf-autofill-controlled-issue-forked-ef3jo?file=/pages/controlled.js:816-950) for a complete example

**Reply by pooyarm** [reply]
yes, thanks. actually I knew it's possible with ref forwarding, but I thought since RHF already keeps those refs, maybe my code could ask RHF for those refs, would be easier and with less code. see:
```jsx
const Input = (props) => {
  const {
    name,
    className,
    defaultValue,
    rules,
    ...restProps
  } = props

  const {
    field: {
      onChange,
      value,
      ref,
      onBlur,
    },
    fieldState: { invalid },
  } = useController({
    name,
    rules,
    defaultValue,
  })

  return (
    <AnotherUILibraryInputComponent
      {...restProps}
      ref={ref}
      onBlur={onBlur}
      onChange={onChange}
      value={value}
      className={classNames(className, 'form__input', {
        'form__input--invalid': invalid,
      })}
    />
  )
}
```

I have a component like above for all my field types, and as I see there is a ref coming from useController, I just thought since RHF has already created the ref, maybe I could use those ones instead of creating a new ref and forwarding through these components deeply.

but anyway, thanks for helping :)

**Reply by Moshyfawn** [reply]
What `useController` exposes is a `RefCallback` to pass the DOM input element into.
<img width="154" alt="image" src="https://user-images.githubusercontent.com/16290753/150118642-44aa0914-f801-45a1-a6d6-79ae4923c577.png">
@bluebill1049, I'm not sure if `useForm` exposes it's collected input `refs` / is it feasible?

**Reply by bluebill1049** [reply]
hey @Moshyfawn we don't expose input reference at the moment, Considering a lot of controlled components doesn't expose any input `ref`, it's not consistent between different UI library.

**Reply by Moshyfawn** [reply]
Fair enough! Thx, @bluebill1049!
